local-volume/provisioner/pkg/discovery/discovery.go

```diff
@@ -181,12 +181,22 @@ func (d *Discoverer) getVolumeType(fullPath string) (string, error) {
 	if isdir {
 		return common.VolumeTypeFile, nil
 	}
+	// check for Block before returning errdir
 	isblk, errblk := d.VolUtil.IsBlock(fullPath)
 	if isblk {
 		return common.VolumeTypeBlock, nil
 	}
 
-	return "", fmt.Errorf("Block device check for %q failed: DirErr - %v BlkErr - %v", fullPath, errdir, errblk)
+	// report the first error found
+	err := errdir
+	if err == nil {
+		err = errblk
+	}
+	if err != nil {
+		return "", fmt.Errorf("Error checking file %q: %s", fullPath, err)
```

Comment:
Can the log message here still refer to the call that generated the error message? Something like this:
if errdir != nil {
   return "", fmt.Errorf("Error checking isDir for path %q: %s", fullPath, err)
}
if errblk != nil {
   return "", fmt.Errorf("Error checking isBlock for path %q: %s", fullPath, err)
}
The extra detail would help with debugging.

Reply:
If you have an errdir, is there still a possibility that it is a block device?  Can you just return immediately if you have an errdir?

Reply:
For some quick tests I just did on CentOS 7 Linux and OS X, it does look like isDir() will not return err just because its a symbolic link to a device. I am trying to recall why I kept error returns independent. I think it might have been because isBlock() was returning errors where it could have been a dir.  In any case, its fine to return immediately after isDir() fails with an err.

Reply:
`IsDir()` returns error only on serious error, like permission issues, file does not exist and such. IsBlock is very likely to return the same error.